charts/x509-certificate-exporter/templates/service.yaml

```diff
@@ -16,10 +16,16 @@ metadata:
     {{- . | toYaml | trim | nindent 4 }}
 {{- end }}
 spec:
+  type: {{ .Values.service.type }}
   ports:
   - name: metrics
     port: {{ .Values.service.port }}
     targetPort: metrics
+    {{- if and .Values.service.nodePort (or (eq .Values.service.type "NodePort") (eq .Values.service.type "LoadBalancer")) }}
```

Comment:
Thanks for the contributing !

I would just simplify a bit:
Proposal #1, simply test nodePort (will be automatically, if ignored)

```
{{- with .Values.service.nodePort }}
nodePort: {{ . }}
{{- end }}
```

Or Proposal #2
```
{{- if and .Values.service.nodePort (not eq .Values.service.type "ClusterIP" }}
nodePort: {{ .Values.service.nodePort }}
{{- end }}
```

Reply:
Agreed that `#1` is sufficient and I like your opinion on letting k8s decide and fail explicitly rather than keeping ignored values. It's nice to have checks and logic to validate inputs but that's better suited for a JSON schema on values, which we could end up adding later.

Reply:
Tested `#1` and it does indeed result in a nice enough error message while keeping the template simple.
```
➤ helm install test . --set service.type=LoadBalancer --set service.nodePort=30123
STATUS: deployed
[...]

➤ helm upgrade test . --set service.nodePort=30123
Error: UPGRADE FAILED: cannot patch "test-x509-certificate-exporter" with kind Service: Service "test-x509-certificate-exporter" is invalid: spec.ports[0].nodePort: Forbidden: may not be used when `type` is 'ClusterIP'
```

Reply:
thanks for the review guys! Prop1 looks good, however question here: what should be the default value of the `nodePort` then? Or just comment it out in the values.yaml all over?

Reply:
Yes leaving `service.nodePort` to it's default value `null` will remove it from the service.
This works well with your patch to `values.yaml` already :+1: :smiley: 
```
➤ helm template test .

# Source: x509-certificate-exporter/templates/service.yaml
apiVersion: v1
kind: Service
metadata:
  name: test-x509-certificate-exporter
  labels:
    helm.sh/chart: x509-certificate-exporter-1.7.0
    app.kubernetes.io/name: x509-certificate-exporter
    app.kubernetes.io/instance: test
    app.kubernetes.io/version: "2.6.1"
    app.kubernetes.io/managed-by: Helm
spec:
  type: ClusterIP
  ports:
  - name: metrics
    port: 9793
    targetPort: metrics
  selector:
    app.kubernetes.io/name: x509-certificate-exporter
    app.kubernetes.io/instance: test
```
```
➤ helm template test . --set service.type=LoadBalancer --set service.nodePort=30123

# Source: x509-certificate-exporter/templates/service.yaml
apiVersion: v1
kind: Service
metadata:
  name: test-x509-certificate-exporter
  labels:
    helm.sh/chart: x509-certificate-exporter-1.7.0
    app.kubernetes.io/name: x509-certificate-exporter
    app.kubernetes.io/instance: test
    app.kubernetes.io/version: "2.6.1"
    app.kubernetes.io/managed-by: Helm
spec:
  type: LoadBalancer
  ports:
  - name: metrics
    port: 9793
    targetPort: metrics
    nodePort: 30123
  selector:
    app.kubernetes.io/name: x509-certificate-exporter
    app.kubernetes.io/instance: test
```